src/certificate.rs

```diff
@@ -180,361 +154,172 @@ pub enum PublicKeyInfo {
-    /// Secure Hash Algorithm 256 (SHA256) and Rivest, Shamir and Adleman (RSA) encryption
-    ///
-    /// See RFC 4055 and RFC 8017.
-    Sha256WithRsaEncryption,
-
-    /// Elliptic Curve Digital Signature Algorithm (DSA) coupled with the Secure Hash
-    /// Algorithm 256 (SHA256) algorithm
-    ///
-    /// See RFC 5758.
-    EcdsaWithSha256,
-}
-
-impl x509::AlgorithmIdentifier for SignatureId {
-    type AlgorithmOid = &'static [u64];
+/// `extensions` is optional; if empty, no extensions will be included.
+#[allow(clippy::too_many_arguments)]
+pub fn generate_self_signed<'a>(
+    yubikey: &mut YubiKey,
+    key: SlotId,
+    serial: &[u8],
+    opt_not_after: Option<Time>,
+    subject: &Name<'a>,
+    spkibuf: &[u8],
+    extensions: Option<Extensions<'a>>,
+    alg_id: AlgorithmId,
+) -> Result<Vec<u8>> {
+    let serial_number = UIntBytes::new(serial).unwrap();
+    let spki = match SubjectPublicKeyInfo::from_der(spkibuf) {
+        Ok(spki) => spki,
+        Err(_e) => return Err(Error::ParseError),
+    };
 
-    fn algorithm(&self) -> Self::AlgorithmOid {
-        match self {
-            SignatureId::Sha256WithRsaEncryption => &[1, 2, 840, 113_549, 1, 1, 11],
-            SignatureId::EcdsaWithSha256 => &[1, 2, 840, 10045, 4, 3, 2],
-        }
+    let mut oid = OID_SHA256_WITH_RSA_ENCRYPTION;
+    if OID_EC_PUBLIC_KEY == spki.algorithm.oid {
+        oid = OID_ECDSA_WITH_SHA256;
+    } else if OID_RSA_ENCRYPTION != spki.algorithm.oid {
+        return Err(Error::AlgorithmError);
     }
 
-    fn parameters<W: std::io::Write>(
-        &self,
-        w: cookie_factory::WriteContext<W>,
-    ) -> cookie_factory::GenResult<W> {
-        // No parameters for any SignatureId
-        Ok(w)
-    }
-}
+    let signature = AlgorithmIdentifier {
+        oid,
+        parameters: None,
+    };
 
-/// Certificates
-#[derive(Clone, Debug)]
-pub struct Certificate {
-    serial: Serial,
-    #[allow(dead_code)]
-    issuer: String,
-    subject: String,
-    subject_pki: PublicKeyInfo,
-    data: Buffer,
-}
+    let ten_years_duration = Duration::from_secs(365 * 24 * 60 * 60 * 10);
+    let ten_years_time = SystemTime::now().checked_add(ten_years_duration).unwrap();
+    let not_after = match opt_not_after {
+        Some(na) => na,
+        None => Time::UtcTime(
+            UtcTime::from_unix_duration(ten_years_time.duration_since(UNIX_EPOCH).unwrap())
+                .unwrap(),
+        ),
+    };
 
-impl<'a> TryFrom<&'a [u8]> for Certificate {
-    type Error = Error;
+    let validity = Validity {
+        not_before: Time::UtcTime(
+            UtcTime::from_unix_duration(SystemTime::now().duration_since(UNIX_EPOCH).unwrap())
+                .unwrap(),
+        ),
+        not_after,
+    };
 
-    fn try_from(bytes: &'a [u8]) -> Result<Self> {
-        Self::from_bytes(bytes.to_vec())
-    }
-}
+    let tbs_certificate = TBSCertificate {
+        version: 2,
```

Comment:
I _thought_ there was a bug here: `version` is set to 2, but `extensions` may be `Some`, and according to `RFC 5280`:

```
issuerUniqueID  [1]  IMPLICIT UniqueIdentifier OPTIONAL,
                     -- If present, version MUST be v2 or v3
subjectUniqueID [2]  IMPLICIT UniqueIdentifier OPTIONAL,
                     -- If present, version MUST be v2 or v3
extensions      [3]  EXPLICIT Extensions OPTIONAL
                     -- If present, version MUST be v3
```

It turns out this is not a bug, because the new `x509` crate doesn't have a type-safe version field, and this value is the raw encoding of the v3 version, not the numeric version v2.

More generally, I dislike this serialization API, enabled as it is by raw-construcing the same struct used as the output of parsing. It is a brittle approach, prone to exactly the kind of mistake I thought was being made here (and could easily be made by someone else trying to use the crate), because no relative constraints are being placed on any of the fields. In particular, I see no handling in the new `x509` crate of the above MUST requirements on `version`; it looks like the user could construct and serialize a certificate that violates them.

Reply:
You are correct there is nothing here that addresses the version requirement (or relative requirements in general). I will give a look at that (easiest way in this case is likely to disallow non-v3 at decode/encode time).

Reply:
FWIW, many of the other crates in https://github.com/rustcrypto/formats use a `Version` enum, e.g:

https://docs.rs/pkcs8/latest/pkcs8/enum.Version.html

It's also computed implicitly from the structure of the document, eliminating the possibility of setting it to the wrong value:

https://docs.rs/pkcs8/latest/pkcs8/struct.PrivateKeyInfo.html#method.version

At decode time, the implicitly computed version is checked against the explicit one in the document to ensure it's set correctly.

Reply:
I will give a look and do something similar (and follow suit in the OCSP and CMS bits as the evolve). Thanks.

Reply:
> We should not be removing the owned types and replacing them in the API with Vec<u8>; this loses significant type safety

There's a "document" type convention for this in the other https://github.com/RustCrypto/formats/ and this PR adds a `CertificateDocument` type: https://github.com/RustCrypto/formats/pull/393

These types are effectively a newtype wrapper around a `Vec<u8>` which is guaranteed to parse successfully as a particular type, i.e. a `Certificate` in this case. They impl the [`der::Document`](https://docs.rs/der/latest/der/trait.Document.html) trait which handles serializing/deserializing DER-encoded documents including PEM (de)serialization as well as reading from and writing to DER or PEM-encoded files.

> More generally, I dislike this serialization API, enabled as it is by raw-construcing the same struct used as the output of parsing. It is a brittle approach, prone to exactly the kind of mistake I thought was being made here (and could easily be made by someone else trying to use the crate), because no relative constraints are being placed on any of the fields.

Regarding this in general, it would be good to add something like a `CertificateBuilder` type which enforces X.509 best practices. We could even use tools like [`certlint`](https://github.com/globalsign/certlint) or [`zlint`](https://github.com/zmap/zlint) to ensure that certificates generated with a `CertificateBuilder` are following those best practices correctly.

The builder type could own the relevant data for various parts of the certificate, and in doing so, that also solves the lifetime-related concerns around a zero-copy certificate parser/serializer which borrows data.

Reply:
@tony-iqlusion @carl-wallace My latest PRs both add `Version` types and make it trivial to derive `Version` types.